compiler/qsc_linter/src/lib.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! This crate contains the linter for the Q# language.
+//!
+//! It includes lints for the following stages of the compilation process:
+//!  - AST
+//!  - HIR
+//!
+//! # Usage
+//!
+//! The entry points to the linter are the `run_*_lints` functions, which take
+//! a `qsc_*::*::Package` as input and output a [`Vec<Lint>`](Lint).
+//!
+//! ## Example
+//!
+//! ```
+//! use linter::run_ast_lints;;
+//! use qsc_ast::ast::Package;
+//!
+//! let package: Package = ...;
+//! let lints: Vec<Lint> = run_ast_lints(&package);
+//! ```
+//!
+//! # How to add a new Lint
+//!
+//! Adding a new lint has three steps:
+//!  1. Declaring the lint: here you set the lint name, the default [`LintLevel`], and the message the user will see.
+//!  2. Implementing the lint: here you write the pattern matching logic of the new lint.
+//!  3. Adding the new lint to the right linter.
+//!
+//! Below is a full example of how to a new AST lint.
+//!
+//! ## Example
+//!
+//! First, we declare and implement our new lint in `src/lints/ast.rs`.
+//! ```
+//! declare_lint!(DoubleParens, LintLevel::Warn, "unnecesary double parentheses")
+//!
+//! // implement the right LintPass for our new lint,
+//! // in this case [`linter::ast::AstLintPass`]
+//! impl linter::ast::AstLintPass for DoubleParens {
+//!     // we only need to impl the relevant check_* method, all the other ones
+//!     // will default to an empty method that will get optmized by rust
+//!     fn check_expr(expr: &qsc_ast::ast::Expr, buffer: &mut Vec<Lint>) {
+//!         // we match the relevant pattern
+//!         if let ExprKind::Paren(ref inner_expr) = *expr.kind {
+//!             if matches!(*inner_expr.kind, ExprKind::Paren(_)) {
+//!                 // we push the lint to the buffer
+//!                 push_lint!(Self, expr, buffer);
+//!             }
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! Finally we add our new lint to `impl CombinedAstLints { ... }` in `src/linter/ast.rs`.
+//!
+//! ```
+//! impl AstLintPass for CombinedAstLints {
+//!     fn check_expr(expr: &Expr) {
+//!         // ... some other lints
+//!
+//!         DoubleParens::check_expr(expr); // add your new lint here
+//!     }
+//! }
+//! ```
+
+#![deny(missing_docs)]
+#![warn(clippy::mod_module_files, clippy::pedantic, clippy::unwrap_used)]
+#![allow(clippy::missing_errors_doc, clippy::missing_panics_doc)]
```

Comment:
we haven't started using the global clippy settings yet (I think that was introduced in 1.76, and our repo currently only required 1.75). So this snippet:
```rust
#![warn(clippy::mod_module_files, clippy::pedantic, clippy::unwrap_used)]
#![allow(clippy::missing_errors_doc, clippy::missing_panics_doc)]
```
Is repeated in every top-level lib.rs file. I'm hoping we can switch to the global settings eventually, but that alone didn't seem like reason to force 1.76. In the meantime, these two lines need to appear at the top of the file to conform to our current pattern of clippy settings.

For the `deny(missing_docs)`, I don't mind trying it out here and then tackling a follow up action to make that a common part of the codebase. One more reason to prefer global settings eventually: applying that more broadly right now will require adding it to each module manually rather than one central location.